### PR TITLE
rpcserver: Support dynamic cert reload.

### PR DESCRIPTION
**Pull request by davecgh**
This modifies the RPC server to support dynamically reloading (aka hot reload) the RPC certificate/key pair as well as the client CAs (when configured with `--authtype=clientcert`).

In other words, dcrd will now notice when the certificates have been changed on the file system on new connections and reload and cache the new ones.

In terms of deciding  when to reload the files, this implementation opts for a highly portable stat-based approach that does not require any additional dependencies over using platform specific file change notifications such as `inotify` on Linux.

This implementation also aims to provide nice error handling semantics and includes additional logic to minimize the amount of disk accesses needed to determine when the files have changed.

The following is an overview of the semantics:

- All connections used a cached TLS config
- Certs are only tested for changes and reloaded when:
  - A new connection is established
  - At least 5 seconds have passed since the last check
  - The file modification times and/or sizes have changed
- The existing working certs are retained if any errors are encountered when loading the new ones in order to avoid breaking a working config
- Only a single error will be shown for attempt at loading an invalid config as opposed to spamming the same error on every new connection

### CODE REVIEW

File: server.go [hunk 1]
```diff
@@ -3269,6 +3269,195 @@ func genCertPair(certFile, keyFile string, altDNSNames []string, tlsCurve ellipt
...
+//
+// An overview of the behavior is as follows:
+//
+//   - All connections used a cached TLS config
+//   - When an underlying file is updated, as determined by its modification
+//     time being newer or its size changing, the certificates are reloaded and
+//     cached
+//   - Files are not checked for updates more than once every several seconds
+//   - Files are only checked for updates when a connection is made and are not
+//     checked more than once ever several seconds
```
Comment (davecgh) [inline server.go:3428]:
```suggestion
//     checked more than once every several seconds
```

Comment (davecgh) [reply server.go]: Done.

File: server.go [hunk 2]
```diff
@@ -3269,6 +3269,195 @@ func genCertPair(certFile, keyFile string, altDNSNames []string, tlsCurve ellipt
...
+	c.nextReloadCheck = now.Add(c.minReloadCheckDelay)
+
+	return c.cert.updated() || c.key.updated() || c.clientCAs.updated()
+}
+
+// newTLSConfig loads the provided server certificate and key pair along with
+// the provided client CAs and returns a new tls.Config instance populated with
+// the parsed values.
+//
+// The clientCAsPath maybe by an empty string when client authentication is not
```
Comment (JoeGruffins) [inline server.go:3348]: maybe by an -> may be an ?

**Review by JoeGruffins — APPROVED** [review]

File: server.go [hunk 3]
```diff
@@ -3269,6 +3269,195 @@ func genCertPair(certFile, keyFile string, altDNSNames []string, tlsCurve ellipt
...
+	// when it very likely was not.
+	fi, err := os.Stat(f.path)
+	if err != nil {
+		return os.IsNotExist(err)
+	}
+	changed := fi.Size() != f.curSize || fi.ModTime() != f.curTime
+	if changed {
+		f.curSize = fi.Size()
+		f.curTime = fi.ModTime()
+	}
```
[the remark this answers is not in the document]
Comment (davecgh) [reply server.go:3306]: Most of the time, yes, but it isn't always necessarily true.

For example, it is quite common on some shared file systems to preserve the modification time when copying files.  Another example is `cp -p` which various sync methods use.

File: server.go [hunk 4]
```diff
@@ -3269,6 +3269,195 @@ func genCertPair(certFile, keyFile string, altDNSNames []string, tlsCurve ellipt
...
 
+// watchedFile houses details about a file that is being watched for updates.
+type watchedFile struct {
+	path    string
+	curTime time.Time
+	curSize int64
+}
+
+// updated returns whether or not the file has been updated since the last time
+// the it was checked and updates the file info details used to make that
```
[the remark this answers is not in the document]
Comment (davecgh) [reply server.go:3280]: Fixed.

File: server.go (hunk 2 above)
Comment (davecgh) [reply server.go]: Updated.

**Review by dajohi — APPROVED** [review]

**Pull request merged**